### PR TITLE
Login service

### PR DESCRIPTION
**Pull request by iomodo**
This PR adds a login service

### CODE REVIEW

File: core/kg/app/user.go [hunk 1]
```diff
@@ -46,6 +50,67 @@ func (a *App) isFirstUserAccount() bool {
 	return count > 0
 }
 
+func (a *App) AuthenticateUserForLogin(userID, username, password string) (*model.User, error) {
+	if userID != "" {
+		user, err := a.store.User().Get(userID)
+		if err != nil {
+			return nil, errors.Wrap(err, "can't get user from store")
+		}
+		return user, nil
```
Comment (bakurits) [inline core/kg/app/user.go:59]: No need to check password when authenticating with ID?

Comment (iomodo) [reply core/kg/app/user.go:59]: good catch @bakurits

File: core/kg/common/utils.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	return res
+}
+
+// GetMillis is a convenience method to get milliseconds since epoch.
+func GetMillis() int64 {
+	return time.Now().UnixNano() / int64(time.Millisecond)
+}
+
+var encoding = base32.NewEncoding("ybndrfg8ejkmcpqxot1uwisza345h769")
```
Comment (giolekva) [inline core/kg/common/utils.go:29]: Where does this string come from?

File: core/kg/common/utils.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package common
```
Comment (giolekva) [inline core/kg/common/utils.go:1]: I strongly discourage use of files like utils/commons and so on.
Over time they tend to accumulate completely unrelated set of functions and becomes hard to read.

Can you split It into separate libs? Sth like: id.go encoding.go time.go ... you decide on exact granularity to split

File: core/kg/common/utils.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/pborman/uuid"
+)
+
+// MapFromJson decodes the key/value pair map
+func MapFromJson(data io.Reader) map[string]string {
+	decoder := json.NewDecoder(data)
+
+	var res map[string]string
+	if err := decoder.Decode(&res); err != nil {
+		return make(map[string]string)
```
Comment (giolekva) [inline core/kg/common/utils.go:19]: Using this in production does not seem safe.
Why swallow/hide the error?

File: core/kg/common/interfaces.go [hunk 5]
```diff
@@ -16,4 +16,12 @@ type AppIface interface {
 	GetUser(userID string) (*model.User, error)
 	CreateUser(user *model.User) (*model.User, error)
 	GetUsers(page, perPage int) ([]*model.User, error)
+
+	AuthenticateUserForLogin(userID, loginID, password string) (*model.User, error)
```
Comment (giolekva) [inline core/kg/common/interfaces.go:20]: In what circumstances would we need to support authenticating users by their userID and not username?
From my understanding userID should be internal to the system.

File: core/kg/app/user.go [hunk 6]
```diff
@@ -59,3 +93,12 @@ func HashPassword(password string) string {
 
 	return string(hash)
 }
+
+// comparePassword compares the hash
+func comparePassword(hash string, password string) bool {
+	if password == "" || hash == "" {
+		return false
+	}
+
```
Comment (giolekva) [inline core/kg/app/user.go:102]: rm extra empty line

File: core/kg/app/user.go [hunk 7]
```diff
@@ -43,7 +43,41 @@ func (a *App) isFirstUserAccount() bool {
...
+		user, err = a.store.User().GetByUsername(username)
+		if err != nil {
+			return nil, errors.Wrapf(err, "can't get user by username")
+		}
+	default:
+		return nil, errors.Wrap(model.ErrInvalidInput, "userID and username are empty")
+	}
+
+	if err := a.checkLogin(user, password); err != nil {
+		return nil, errors.Wrapf(err, "login error")
```
Comment (giolekva) [inline core/kg/app/user.go:68]: "login error" seams too generic, how about using "invalid credentials" ?

File: core/kg/model/role.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+	return r.IsValidWithoutID()
+}
+
+func (r *Role) IsValidWithoutID() bool {
+	if r.Name == "" || len(r.Name) > RoleNameMaxLength {
+		return false
+	}
+
+	if strings.TrimLeft(r.Name, "abcdefghijklmnopqrstuvwxyz0123456789_") != "" {
```
Comment (giolekva) [inline core/kg/model/role.go:33]: Create a RoleNameAllowedChars const string from "abide...." ?

File: core/kg/model/role.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	return r.IsValidWithoutID()
+}
+
+func (r *Role) IsValidWithoutID() bool {
+	if r.Name == "" || len(r.Name) > RoleNameMaxLength {
+		return false
+	}
+
```
Comment (giolekva) [inline core/kg/model/role.go:32]: rm extra empty lines, here and bellow

File: core/kg/model/permission.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+var PermissionGetUsers *Permission
+
+func initializePermissions() {
+	PermissionGetUsers = &Permission{
+		ID:          "get_users",
+		Name:        "Get Users",
+		Description: "gets user list",
+	}
+
```
Comment (giolekva) [inline core/kg/model/permission.go:19]: rm empty line

File: core/kg/model/role.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	if strings.TrimLeft(r.Name, "abcdefghijklmnopqrstuvwxyz0123456789_") != "" {
+		return false
+	}
+
+	if len(r.Description) > RoleDescriptionMaxLength {
+		return false
+	}
+
+	check := func(perms []*Permission, permission string) bool {
```
Comment (giolekva) [inline core/kg/model/role.go:41]: How about creating AllPermissionIds const field next to AllPermissions in permission.go so we don't have to do a linear search here?

File: core/kg/model/errors.go [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package model
+
+import "errors"
+
+var (
+	ErrNotFound     = errors.New("not found")
+	ErrInvalidInput = errors.New("invalid input")
+	ErrUnauthorized = errors.New("access denied")
+	ErrForbidden    = errors.New("forbidden")
```
Comment (giolekva) [inline core/kg/model/errors.go:9]: What is the difference between access denied and forbidden?

File: core/kg/store/sqlstore/session_store.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+	schema := `CREATE TABLE IF NOT EXISTS Sessions (
+			id VARCHAR(26) PRIMARY KEY,
+			token VARCHAR(26), 
+			create_at INTEGER,
+			expires_at INTEGER,
+			last_activity_at INTEGER,
+			user_id VARCHAR(26);`
+
+	ss.db.MustExec(schema)
```
Comment (giolekva) [inline core/kg/store/sqlstore/session_store.go:36]: Does squirrel always run CREATE TABLE IF ... on every query?
First time I'm seeing such pattern

**Review by giolekva — CHANGES_REQUESTED** [review]
Unit tests? :(

**Pull request closed without merging**